src/main/java/ApplyToGroup/ApplyToGroupInteractor.java

```diff
@@ -0,0 +1,58 @@
+package ApplyToGroup;
+import Entities.User;
+import Entities.Group;
+import Entities.CurrentUser;
+
+public class ApplyToGroupInteractor implements ApplyToGroupInputBoundary {
+
+    final ApplyToGroupDsGateway applyToGroupDsGateway;
+    final ApplyToGroupPresenter applyToGroupPresenter;
+    final User user = CurrentUser.getUser();
+
+    public ApplyToGroupInteractor(ApplyToGroupDsGateway applyToGroupDsGateway,
+                                ApplyToGroupPresenter applyToGroupPresenter) {
+
+        this.applyToGroupDsGateway = applyToGroupDsGateway;
+        this.applyToGroupPresenter = applyToGroupPresenter;
+    }
+
+    @Override
+    public ApplyToGroupResponseModel create(ApplyToGroupRequestModel requestModel) {
+        if (!applyToGroupDsGateway.userExistsByName(user.getUsername())) {
+            return applyToGroupPresenter.prepareFailView("User does not exist.");
+        }
+        if (!applyToGroupDsGateway.groupExistsByName(requestModel.getGroupName())) {
+            return applyToGroupPresenter.prepareFailView("Group does not exist.");
+        }
+
+        Group group = applyToGroupDsGateway.getGroup(requestModel.getGroupName());
+
+        if (group.getGroupMembers().containsValue(user)) {
+            return applyToGroupPresenter.prepareFailView("User is already in group.");
+        }
+
+        if (group.getMemberRequests().containsValue(user)) {
+            return applyToGroupPresenter.prepareFailView("User has already applied to the group.");
+        }
+
+        if (user.getGroups().containsValue(group)) {
+            return applyToGroupPresenter.prepareFailView("Group is already in user's list.");
+        }
+
+        if (user.getApplicationsList().containsValue(group)) {
+            return applyToGroupPresenter.prepareFailView("User has already applied to the group.");
+        }
+
+        if (group.getGroupLeader().getUsername().equals(user.getUsername())) {
+            return applyToGroupPresenter.prepareFailView("User is group leader.");
```

Comment:
Some of these checks seem trivial.. please correct me if I'm missing something. It seems you performed the same checks that I did in my usecase, but our usecases are not THAT similar. How would we ever get a situation where a user can click the "apply to group" button if the user is already in that group or its group leader? The view or screen should never be in that position and displaying that button. Let's get together and discuss this before our next meeting. This is the line of thought that made me delete some of my own checks too.

Reply:
I wanted to make sure that I didn't miss anything but it also does not make sense to me. I think I just did not have enough information or understanding of the screens yet so put any checks I could with the intention of deleting it later.